src/libs/safe-path/src/safe_dir_builder.rs

```diff
@@ -0,0 +1,162 @@
+// Copyright (c) 2022 Alibaba Cloud
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use std::fs::DirBuilder;
+use std::io::{Error, ErrorKind, Result};
+use std::os::unix::fs::DirBuilderExt;
+use std::path::{Path, PathBuf};
+
+use crate::{safe_join, SafePathBuf};
+
+/// Safe version of `DirBuilder` to protect from TOCTOU style of attacks.
+#[derive(Debug)]
+pub struct SafeDirBuilder {
+    root: PathBuf,
+    mode: u32,
+    recursive: bool,
+}
+
+impl SafeDirBuilder {
+    /// Creates a new set of options with default mode/security settings for all platforms and
+    /// also non-recursive.
+    pub fn new() -> Self {
+        SafeDirBuilder {
+            root: PathBuf::from("/"),
+            mode: 0o777,
+            recursive: false,
+        }
+    }
+
+    /// Indicates that directories should be created recursively, creating all parent directories.
+    /// Parents that do not exist are created with the same security and permissions settings.
+    pub fn recursive(&mut self, recursive: bool) -> &mut Self {
+        self.recursive = recursive;
+        self
+    }
+
+    /// Sets the mode to create new directories with. This option defaults to 0o777.
+    pub fn mode(&mut self, mode: u32) -> &mut Self {
+        self.mode = mode;
+        self
+    }
+
+    /// Sets the root directory for security check.
+    ///
+    /// # Safety
+    /// The `root` is the trusted part, and should not be modifiable by untrusted agents.
+    pub fn root<P: AsRef<Path>>(&mut self, root: P) -> Result<&mut Self> {
```

Comment:
Since setting `root` is essential to the safe operation of `SafeDirBuilder`, would it make more sense to give that as an argument to `new()` rather than setting it afterwards?

Reply:
Fixed